ui/FilesPanel.go

```diff
@@ -90,33 +90,58 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
 
 func (this *filesPanel) createBackButton() *gtk.Button {
 	image := utils.MustImageFromFileWithSize("back.svg", this.Scaled(40), this.Scaled(40))
-	return utils.MustButton(image, func() {
-		if this.locationHistory.Length() < 1 {
-			this.UI.GoToPreviousPanel()
-		} else if this.locationHistory.IsRoot() {
-			this.locationHistory.GoBack()
-			this.doLoadFiles()
-		} else {
-			this.locationHistory.GoBack()
-			this.doLoadFiles()
-		}
-	})
+	return utils.MustButton(image, this.goBack)
 }
 
 func (this *filesPanel) doLoadFiles() {
 	utils.EmptyTheContainer(&this.listBox.Container)
-
-	if this.displayRootLocations() {
-		this.addRootLocations()
-	} else {
+	if this.isRoot() {
+		if this.refreshSD() {
+			this.addRootLocations()
+		} else {
+			this.locationHistory = utils.LocationHistory {
+				Locations: []dataModels.Location{dataModels.Local},
+			}
+		}
+	}
+	if !this.isRoot() {
 		sortedFiles := this.getSortedFiles()
 		this.addSortedFiles(sortedFiles)
 	}
-
 	this.listBox.ShowAll()
 }
 
-func (this *filesPanel) displayRootLocations() bool {
+func (this *filesPanel) refreshSD() bool {
+	err := (&octoprintApis.SdRefreshRequest {}).Do(this.UI.Client)
+	if err == nil {
+		sdState, err := (&octoprintApis.SdStateRequest {}).Do(this.UI.Client)
+		if err == nil && sdState.IsReady == true {
+			return true
+		} else {
+			return false
+		}
+	} else {
+		return false
+	}
+}
+
+func (this *filesPanel) goBack() {
+	if this.isRoot() {
+		this.UI.GoToPreviousPanel()
+	} else if this.locationHistory.IsRoot() {
+		this.locationHistory.GoBack()
+		if this.refreshSD() {
```

Comment:
It does occur to me that this will cause `refreshSD` to trigger twice (here, as well as in `doLoadFiles`), however there doesn't seem to be any kind of performance penalty from this that I can detect.

Reply:
I was going to suggest passing a bool into doLoadFiles(), but it's used as a callback and so it can't be passed in.  It makes an extra REST call, which is expensive.  The next thing which comes to mind is adding a class member, set it before calling doLoadFiles() and then reset it, but I really don't like this as it really pollutes the class for just this one edge case.  Let's leave it in for now.

Reply:
I actually did some testing, and it doesn't actually add any perceptible overhead. The menu doesn't hesitate in any noticeable way. Further, doing the call in both directions (and subsequently possibly twice) does account for inserting an SD at any time.

Reply:
@JeffB42 - So I just ran this side-by-side with another branch with the same feature, there may have been a fraction of a second hesitation, but it didn't really impact the responsiveness or UX.